File: Data Structures & Algorithms/interleaving-string/submission-23.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:

    # i   0   1   2   3           j   0   1   2   3
    #     a   a   a   a               b   b   b   b

    # k   0   1   2   3   4   5   6   7
    #     a   a   b   b   b   b   a   a
        M,N,O= len(s1),len(s2),len(s3)
        if M+N != O:
            return False

        dp = [False]*(N+1)
        dp[N] = True
        for i in range(M,-1,-1):
            cur = [False]*(N+1)
            if i ==M:
                cur[N] = True
            for j in range(N,-1,-1):
                if cur[j]:
                    continue
                k = i+j
                if i<M and s1[i] == s3[k]:
                    cur[j] |= dp[j]
                if j<N and s2[j] == s3[k]:
                    cur[j] |= cur[j+1]
            dp = cur
        return dp[0]

        # def dfs(i,j):
        #     k =i+j
        #     if k == O:
        #         return True
        #     if (i,j) in dp:
        #         return dp[(i,j)]
        #     if i<M and s1[i]== s3[k] and dfs(i+1,j):
        #         return True
        #     if j<N and s2[j] == s3[k] and dfs(i,j+1):
        #         return True
        #     dp[(i,j)] = False
        #     return False
        # return dfs(0,0)




        

File: Data Structures & Algorithms/longest-common-subsequence/submission-23.java
class Solution {
    int[][] dp;
    public int longestCommonSubsequence(String text1, String text2) {
        dp = new int[text1.length()+1][text2.length()+1];
        for(int i1=text1.length()-1;i1>=0;i1--){
            for(int i2=text2.length()-1;i2>=0;i2--){
                if(text1.charAt(i1)!= text2.charAt(i2)){
                    dp[i1][i2] = Math.max(dp[i1][i2+1],dp[i1+1][i2]);
                } else {
                    dp[i1][i2] = 1+dp[i1+1][i2+1];
                }
            }
        }
        return dp[0][0];
        // return dfs(0,0,text1,text2);
    }

    // public int dfs(int i1,int i2, String text1,String text2){
    //     if(i1 == text1.length() || i2 == text2.length()){
    //         return 0;
    //     }
    //     if(dp[i1][i2] != null) {
    //         return dp[i1][i2];
    //     }
    //     if(text1.charAt(i1)!= text2.charAt(i2)){
    //         dp[i1][i2] = Math.max(dfs(i1+1,i2,text1,text2),dfs(i1,i2+1,text1,text2));
    //     } else {
    //         dp[i1][i2] = 1 + dfs(i1+1,i2+1,text1,text2);
    //     }
    //     return dp[i1][i2];
    // }
}
